axum/src/extract/matched_path.rs

```diff
@@ -49,10 +53,47 @@ use std::sync::Arc;
 ///             tracing::info_span!("http-request", %path)
 ///         }),
 ///     );
-/// # async {
-/// # axum::Server::bind(&"".parse().unwrap()).serve(app.into_make_service()).await.unwrap();
-/// # };
+/// # let _: Router = app;
 /// ```
+///
+/// # Matched path in nested routers
+///
+/// Because of how [nesting] works `MatchedPath` isn't accessible in middleware on nested routes:
+///
+/// ```
+/// use axum::{
+///     Router,
+///     RequestExt,
+///     routing::get,
+///     extract::{MatchedPath, rejection::MatchedPathRejection},
+///     middleware::map_request,
+///     http::Request,
+///     body::Body,
+/// };
+///
+/// async fn access_matched_path(mut request: Request<Body>) -> Request<Body> {
+///     // if `/foo/bar` is called this will be `Err(_)` since that matches
+///     // a nested route
+///     let matched_path: Result<MatchedPath, MatchedPathRejection> =
+///         request.extract_parts::<MatchedPath>().await;
+///
+///     request
+/// }
+///
+/// // `MatchedPath` is always accessible on handlers regardless
+/// // if its for a nested route or not
```

Comment:
```suggestion
/// // of whether it's for a nested route or not
```

Reply:
Also, what if you do `.nest("/foo", handler.into_service())`? Doesn't count as nesting a handler? 😄

Reply:
Oh right good point. Such handlers should be able to access the matched path either. I've added a test case for it and clarified the docs a bit.